pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -621,30 +621,88 @@ func translateFault(in *networking.HTTPFaultInjection) *xdshttpfault.HTTPFault {
 	return &out
 }
 
-func getHashPolicy(configStore model.IstioConfigStore, hostname model.Hostname) *route.RouteAction_HashPolicy {
+func portLevelSettingsConsistentHash(dst *networking.Destination,
+	pls []*networking.TrafficPolicy_PortTrafficPolicy) *networking.LoadBalancerSettings_ConsistentHashLB {
+	if dst.Port != nil {
+		switch dst.Port.Port.(type) {
+		case *networking.PortSelector_Name:
+			log.Warnf("using deprecated name on port selector - ignoring")
+		case *networking.PortSelector_Number:
+			portNumber := dst.GetPort().GetNumber()
+			for _, setting := range pls {
+				number := setting.GetPort().GetNumber()
+				if number == portNumber {
+					return setting.GetLoadBalancer().GetConsistentHash()
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func getHashPolicy(configStore model.IstioConfigStore, dst *networking.DestinationWeight) *route.RouteAction_HashPolicy {
 	if configStore == nil {
 		return nil
 	}
 
-	destinationRule := configStore.DestinationRule(hostname)
+	destination := dst.GetDestination()
+	destinationRule := configStore.DestinationRule(model.Hostname(destination.GetHost()))
 	if destinationRule == nil {
 		return nil
 	}
-
 	rule := destinationRule.Spec.(*networking.DestinationRule)
+
 	consistentHash := rule.GetTrafficPolicy().GetLoadBalancer().GetConsistentHash()
+
+	for _, subset := range rule.GetSubsets() {
+		if subset.GetName() == destination.GetSubset() {
+			subsetPortLevelSettings := subset.GetTrafficPolicy().GetPortLevelSettings()
+			consistentHash = subset.GetTrafficPolicy().GetLoadBalancer().GetConsistentHash()
+			settingsHash := portLevelSettingsConsistentHash(destination, subsetPortLevelSettings)
+			if settingsHash != nil {
+				consistentHash = settingsHash
+			}
```

Comment:
No if condition here. Essentially there is no inheritance. If something is missing it goes to default values. So you can assign everything to consistentHash and do a nil check in the end.

Reply:
We'd be able to move the nil checks if this were patterned more after `cluster.go`. We'll see about clarifying the behavior so it's more obvious in the next commit. We were hoping to avoid modifying the hash policy in place for clarity's sake.